### PR TITLE
DAOS-8185 client: Improve error message if agent not running.

### PR DESCRIPTION
**Pull request by ashleypittman**
Use actual error number rather than incorrect -DER_MISC if
the agent isn't found.

The code here was incorrect in it was passing -1 to errno2der()
which was returning -DER_MISC, fix the client code so that it
returns -DER_NONEXISTS correctly, and also change errno2der()
to use -DER_UNKNOWN for negative error numbers.

Update other error paths in connect code to use correct
errors rather than -DER_MISC.

Signed-off-by: Ashley Pittman <ashley.m.pittman@intel.com>

### CODE REVIEW

File: src/include/daos/common.h [hunk 1]
```diff
@@ -495,6 +496,11 @@ void daos_iov_append(d_iov_t *iov, void *buf, uint64_t buf_len);
 static inline int
 daos_errno2der(int err)
 {
+	if (err < 0) {
+		D_ERROR("error < 0 (%d)\n", err);
+		return -DER_UNKNOWN;
+	}
+
```
Comment (ashleypittman) [inline src/include/daos/common.h:503]: With these two checks I did wonder about leaving them out of release builds as it's basically a sanity check.

**Review by mjmac — APPROVED** [review]

**Comment by ashleypittman** [comment]
I spoke with Johann about this who said to bring in @chrisphoffman

I think this PR is correct, and that it fixes a bug, but it potentially also means that daos is returning -DER_NOENT if either the agent isn't running or a pool doesn't exist and whilst that is what master intends to do is probably not desirable, so we probably want to change the behaviour of agent connect in this case to return something else specific.

**Comment by chrisphoffman** [comment]
> 
> 
> I spoke with Johann about this who said to bring in @chrisphoffman
> 
> I think this PR is correct, and that it fixes a bug, but it potentially also means that daos is returning -DER_NOENT if either the agent isn't running or a pool doesn't exist and whilst that is what master intends to do is probably not desirable, so we probably want to change the behaviour of agent connect in this case to return something else specific.

I've been working on DAOS-7487. One of the requirements is to improve error reporting and improve checking connectivity, where needed, to server for autotest. This has two pieces one for client to agent and then also from agent to server.

In regards to client to agent we had discussion around using an errono that covered a failure of agent not running/unreachable. If a suitable one wasn't found one was to be created. DER_AGENT or something similar was discussed. 

The question is DER_NOENT specific enough for the failure or do we need something specific for this type of failure?

**Comment by ashleypittman** [comment]
> > I spoke with Johann about this who said to bring in @chrisphoffman
> > I think this PR is correct, and that it fixes a bug, but it potentially also means that daos is returning -DER_NOENT if either the agent isn't running or a pool doesn't exist and whilst that is what master intends to do is probably not desirable, so we probably want to change the behaviour of agent connect in this case to return something else specific.
> 
> I've been working on DAOS-7487. One of the requirements is to improve error reporting and improve checking connectivity, where needed, to server for autotest. This has two pieces one for client to agent and then also from agent to server.
> 
> In regards to client to agent we had discussion around using an errono that covered a failure of agent not running/unreachable. If a suitable one wasn't found one was to be created. DER_AGENT or something similar was discussed.
> 
> The question is DER_NOENT specific enough for the failure or do we need something specific for this type of failure?

I think that DER_NOENT also means the pool doesn't exist, so unless there's a way of distinguishing those two failures then I think we need something else.

**Review by chrisphoffman — APPROVED** [review]

**Comment by chrisphoffman** [comment]
> 
> 
> > > I spoke with Johann about this who said to bring in @chrisphoffman
> > > I think this PR is correct, and that it fixes a bug, but it potentially also means that daos is returning -DER_NOENT if either the agent isn't running or a pool doesn't exist and whilst that is what master intends to do is probably not desirable, so we probably want to change the behaviour of agent connect in this case to return something else specific.
> > 
> > 
> > I've been working on DAOS-7487. One of the requirements is to improve error reporting and improve checking connectivity, where needed, to server for autotest. This has two pieces one for client to agent and then also from agent to server.
> > In regards to client to agent we had discussion around using an errono that covered a failure of agent not running/unreachable. If a suitable one wasn't found one was to be created. DER_AGENT or something similar was discussed.
> > The question is DER_NOENT specific enough for the failure or do we need something specific for this type of failure?
> 
> I think that DER_NOENT also means the pool doesn't exist, so unless there's a way of distinguishing those two failures then I think we need something else.

Spoke with Ashley, DER_AGENT will be implemented in the mentioned ticket.

**Review by johannlombardi — APPROVED** [review]

**Pull request merged**